File: services/api/app/core/contracts.py
from enum import Enum
from typing import Literal, Protocol
from pydantic import BaseModel, Field

# ---------- errors ----------
class MissingAPIKeyError(RuntimeError): ...       # HR-2 — raised at startup, never caught to degrade
class ProviderRateLimited(RuntimeError): ...
class ParseFailure(RuntimeError): ...
class KernelRejection(RuntimeError): ...

# ---------- sources ----------
class AbstractSource(str, Enum):
    S2 = "s2"; OPENALEX_INVERTED = "openalex_inverted"; TLDR = "tldr"; UNAVAILABLE = "unavailable"

class Provenance(BaseModel):
    provider: Literal["semantic_scholar", "openalex", "crossref"]
    endpoint: str
    retrieved_at: str
    external_url: str                              # must be a real, resolvable URL

class SourceRecord(BaseModel):
    source_id: str
    csl: dict                                      # CSL-JSON — the one canonical citation model
    provenance: Provenance                         # HR-1: proof this came from an HTTP response
    abstract: str | None = None
    abstract_source: AbstractSource = AbstractSource.UNAVAILABLE

class SourceStore(Protocol):
    """APPEND-ONLY. Only app/providers/* may call put(). HR-1."""
    def put(self, record: SourceRecord) -> str: ...
    def get(self, source_id: str) -> SourceRecord | None: ...
    def has(self, source_id: str) -> bool: ...

# ---------- document IR ----------
class CitationAnchor(BaseModel):
    anchor_id: str
    source_ids: list[str]                          # FK into SourceStore — validated by the kernel
    offset_in_span: int
    original_marker_text: str | None = None
    provenance_kind: Literal["parsed", "agent_added"] = "parsed"
    confidence: float = 1.0
    fingerprint_id: str | None = None              # FK → anchor_fingerprints table (ADR-017).
                                                   # Vectors are NEVER stored inline: they would
                                                   # duplicate per version and make structural
                                                   # diffs unreadable.
    locator: str | None = None
    prefix: str | None = None

class Span(BaseModel):
    id: str
    text: str                                      # text lives ONLY here
    citation_anchors: list[CitationAnchor] = Field(default_factory=list)

class Block(BaseModel):
    id: str
    type: Literal["paragraph", "equation", "figure", "table", "list"]
    order: int
    spans: list[Span] = Field(default_factory=list)
    placeholder_caption: str | None = None         # figures/tables/equations: caption only (ADR-008)

class Section(BaseModel):
    id: str; level: int; title: str; order: int
    blocks: list[Block] = Field(default_factory=list)

class QuarantineEntry(BaseModel):
    raw: str
    reason: Literal["parse_failed", "unresolved", "orphan_marker", "segmentation_failed"]
    page: int | None = None

class DocumentMeta(BaseModel):
    title: str | None = None
    style_id: str | None = None
    style_confidence: float | None = None
    style_ambiguous: bool = False

class Document(BaseModel):
    doc_id: str
    version: int
    metadata: DocumentMeta
    sections: list[Section] = Field(default_factory=list)
    quarantine: list[QuarantineEntry] = Field(default_factory=list)

# ---------- parsing ----------
class ConfidenceTier(str, Enum):
    RESOLVED = "resolved"; PARSED_UNRESOLVED = "parsed_unresolved"
    LOW_CONFIDENCE = "low_confidence"; QUARANTINED = "quarantined"
    ORPHAN_MARKER = "orphan_marker"

class ParsedReference(BaseModel):
    ref_id: str
    raw_string: str                                # always retained, verbatim
    csl: dict | None
    tier: ConfidenceTier
    parse_confidence: float
    agreement_score: float | None = None           # arbiter; accept at >= 0.85
    source_id: str | None = None

# ---------- review ----------
class Claim(BaseModel):
    claim_id: str; text: str; span_id: str
    anchor_ids: list[str] = Field(default_factory=list)
    citability: float                              # streaming order = descending citability

class Candidate(BaseModel):
    source_id: str
    strategy: Literal["s2_snippet", "s2_recommendations", "openalex_search", "openalex_graph"]
    fused_score: float
    rerank_score: float | None = None

class VerificationLabel(str, Enum):
    SUPPORTS = "supports"; PARTIALLY_SUPPORTS = "partially_supports"
    DOES_NOT_ADDRESS = "does_not_address"; CONTRADICTS = "contradicts"
    UNVERIFIABLE_NO_ABSTRACT = "unverifiable_no_abstract"

class Verification(BaseModel):
    label: VerificationLabel
    quote: str | None                              # MUST be a substring of the fetched abstract
    abstract_source: AbstractSource
    confidence: float

class Finding(BaseModel):
    finding_id: str
    kind: Literal["missing_work", "claim_citation_mismatch", "no_candidates_found"]
    claim: Claim
    source_id: str | None
    verification: Verification | None
    severity: Literal["high", "medium", "low", "info"]

# ---------- agent ----------
class OperationType(str, Enum):
    ADD_CITATIONS = "AddCitations"; FIND_SUPPORT = "FindSupport"
    SHORTEN = "Shorten"; REWRITE_SECTION = "RewriteSection"
    REPLACE_CITATION = "ReplaceCitation"; MOVE_TEXT = "MoveText"
    FREEFORM_EDIT = "FreeformEdit"

class Operation(BaseModel):
    op: OperationType
    target_ids: list[str]
    params: dict = Field(default_factory=dict)
    no_typed_op_applies: bool = False              # required True for FREEFORM_EDIT
    justification: str | None = None               # required for FREEFORM_EDIT

class EditPlan(BaseModel):
    plan_id: str
    operations: list[Operation]

class ProposedChange(BaseModel):
    change_id: str
    op: Operation
    new_fragment: dict                             # partial IR
    new_source_ids: list[str] = Field(default_factory=list)
    orphaned_anchor_ids: list[str] = Field(default_factory=list)
    rationale: str

class KernelVerdict(BaseModel):
    decision: Literal["accept", "reject", "flag"]
    reasons: list[str]                             # never empty for reject/flag
    flags: list[str] = Field(default_factory=list)

# ---------- providers ----------
class Provider(Protocol):
    """Implementations whose API degrades silently without credentials MUST raise
    MissingAPIKeyError at construction when theirs is absent (HR-2). One whose API
    throttles with an error status we already raise on MAY run unauthenticated —
    see ADR-010a, and read it before adding a second exception."""
    async def search_works(self, query: str, limit: int = 10) -> list[SourceRecord]: ...
    async def match_reference(self, title: str, year: int | None = None) -> SourceRecord | None: ...
    async def get_abstract(self, source_id: str) -> tuple[str | None, AbstractSource]: ...
    async def batch_hydrate(self, ids: list[str]) -> list[SourceRecord]: ...

# ---------- LLM (ADR-015 / 016 / 018) ----------
class LLMRole(str, Enum):
    """Model is chosen per ROLE, never globally. IDs pinned in config.py — no model
    string appears anywhere else in the codebase."""
    REPAIR = "repair"                  # gpt-5.4-mini  — reference segment-and-label
    CLAIM_EXTRACTION = "claim_extraction"  # gpt-5.4
    RERANK = "rerank"                  # gpt-5.4-mini  — after the embedding prefilter
    VERIFY = "verify"                  # gpt-5.5       — accuracy-critical, do not economise
    PLAN = "plan"                      # gpt-5.5
    TRANSFORM = "transform"            # gpt-5.4       — rewriting the user's prose
    ORCHESTRATE = "orchestrate"        # gpt-5.5       — tool-call routing over a conversation

class LLMClient(Protocol):
    """The ONLY path to OpenAI. Structured output is mandatory for every data-returning
    call — JSON Schema, not prompt-and-parse. Honours LLM_MODE=record|replay|live;
    in replay a cache miss RAISES rather than calling the API (ADR-018)."""
    async def complete(
        self, role: LLMRole, prompt: str, schema: dict, *, system: str | None = None
    ) -> dict: ...
    async def embed(self, texts: list[str]) -> list[list[float]]: ...   # 3-small @ 512 dims

class JobStatus(str, Enum):
    QUEUED = "queued"; RUNNING = "running"; SUCCEEDED = "succeeded"; FAILED = "failed"

class Job(BaseModel):
    """A crashed worker is a FAILURE and must be visible. A UI streaming nothing forever
    reads as 'no findings' — the same false negative as ADR-010. HR-3."""
    job_id: str
    kind: Literal["ingest", "review"]
    status: JobStatus
    progress_current: int = 0
    progress_total: int = 0
    error: str | None = None

File: apps/web/src/lib/contracts.ts
/**
 * TypeScript transcription of goal.md Appendix A — the frozen contracts.
 *
 * This file mirrors `services/api/app/core/contracts.py`. It is authoritative
 * for the frontend: every screen is built against these shapes, whether the
 * data comes from the live API or from the typed fixtures in lib/api/fixtures.
 *
 * Do not "improve" a shape here. If the frontend needs something these types
 * cannot express, that is an Interface Request in memory.md §5, not an edit.
 */

// ---------- sources ----------
export type AbstractSource = 's2' | 'openalex_inverted' | 'tldr' | 'unavailable';

export interface Provenance {
  provider: 'semantic_scholar' | 'openalex' | 'crossref';
  endpoint: string;
  retrieved_at: string;
  /** HR-1: must be a real, resolvable URL. The UI always links it. */
  external_url: string;
}

export interface SourceRecord {
  source_id: string;
  /** CSL-JSON — the one canonical citation model (HR-4). Never pre-formatted. */
  csl: CslJson;
  provenance: Provenance;
  abstract?: string | null;
  abstract_source: AbstractSource;
}

/** Minimal structural view of CSL-JSON. citation.js consumes the whole object. */
export interface CslJson {
  id?: string;
  type?: string;
  title?: string;
  author?: Array<{ family?: string; given?: string; literal?: string }>;
  issued?: { 'date-parts'?: number[][]; raw?: string };
  'container-title'?: string;
  volume?: string;
  issue?: string;
  page?: string;
  DOI?: string;
  URL?: string;
  publisher?: string;
  [key: string]: unknown;
}

// ---------- document IR ----------
export interface CitationAnchor {
  anchor_id: string;
  /** FK into SourceStore — validated by the kernel. */
  source_ids: string[];
  offset_in_span: number;
  original_marker_text?: string | null;
  provenance_kind: 'parsed' | 'agent_added';
  confidence: number;
  locator?: string | null;
  prefix?: string | null;
}

export interface Span {
  id: string;
  /** Text lives ONLY here. */
  text: string;
  citation_anchors: CitationAnchor[];
}

export type BlockType = 'paragraph' | 'equation' | 'figure' | 'table' | 'list';

export interface Block {
  id: string;
  type: BlockType;
  order: number;
  spans: Span[];
  /** figures/tables/equations: caption only (ADR-008). */
  placeholder_caption?: string | null;
}

export interface Section {
  id: string;
  level: number;
  title: string;
  order: number;
  blocks: Block[];
}

export type QuarantineReason =
  | 'parse_failed'
  | 'unresolved'
  | 'orphan_marker'
  | 'segmentation_failed';

export interface QuarantineEntry {
  raw: string;
  reason: QuarantineReason;
  page?: number | null;
}

export interface DocumentMeta {
  title?: string | null;
  style_id?: string | null;
  style_confidence?: number | null;
  style_ambiguous: boolean;
}

export interface DocumentIR {
  doc_id: string;
  version: number;
  metadata: DocumentMeta;
  sections: Section[];
  quarantine: QuarantineEntry[];
}

// ---------- parsing ----------
export type ConfidenceTier =
  | 'resolved'
  | 'parsed_unresolved'
  | 'low_confidence'
  | 'quarantined'
  | 'orphan_marker';

export interface ParsedReference {
  ref_id: string;
  /** Always retained, verbatim. Rendered in mono, never truncated. */
  raw_string: string;
  csl: CslJson | null;
  tier: ConfidenceTier;
  parse_confidence: number;
  /** Arbiter; accepted at >= 0.85. */
  agreement_score?: number | null;
  source_id?: string | null;
}

/** An in-text marker whose target reference does not exist (tier orphan_marker). */
/**
 * An in-text marker whose bibliography entry does not exist.
 *
 * There is no `ref_id` here on purpose: an orphan marker is defined by having no
 * reference, so `anchor_id` is its identity. This type used to declare `ref_id` and a
 * non-optional `snippet`, and the backend sent neither — `snippet.split()` then threw
 * and took the whole inspector down on the first paper that actually had one.
 */
export interface OrphanMarker {
  anchor_id: string;
  marker_text: string;
  span_id: string;
  section_id: string;
  target: string | null;
  reason: string;
  page: number | null;
  /** The sentence the marker sits in, resolved from the IR. Null when it could not
   *  be located — stated rather than substituted. */
  snippet: string | null;
  section_title: string | null;
}

// ---------- review ----------
export interface Claim {
  claim_id: string;
  text: string;
  span_id: string;
  anchor_ids: string[];
  /** Streaming order = descending citability. */
  citability: number;
}

export type VerificationLabel =
  | 'supports'
  | 'partially_supports'
  | 'does_not_address'
  | 'contradicts'
  | 'unverifiable_no_abstract';

export interface Verification {
  label: VerificationLabel;
  /** MUST be a verbatim substring of the fetched abstract (ADR-006). */
  quote: string | null;
  abstract_source: AbstractSource;
  confidence: number;
}

export type FindingKind = 'missing_work' | 'claim_citation_mismatch' | 'no_candidates_found';

export interface Finding {
  finding_id: string;
  kind: FindingKind;
  claim: Claim;
  source_id: string | null;
  verification: Verification | null;
  severity: 'high' | 'medium' | 'low' | 'info';
}

// ---------- agent ----------
export type OperationType =
  | 'AddCitations'
  | 'FindSupport'
  | 'Shorten'
  | 'RewriteSection'
  | 'ReplaceCitation'
  | 'MoveText'
  | 'FreeformEdit';

export interface Operation {
  op: OperationType;
  target_ids: string[];
  params: Record<string, unknown>;
  /** Required true for FreeformEdit (ADR-009). */
  no_typed_op_applies: boolean;
  justification?: string | null;
}

export interface ProposedChange {
  change_id: string;
  op: Operation;
  /** Partial IR. */
  new_fragment: Record<string, unknown>;
  new_source_ids: string[];
  orphaned_anchor_ids: string[];
  rationale: string;
}

export interface KernelVerdict {
  decision: 'accept' | 'reject' | 'flag';
  /** Never empty for reject/flag. */
  reasons: string[];
  flags: string[];
}
